src/verticalaligner.cpp

```diff
@@ -151,6 +187,54 @@ void StaffAlignment::SetStaff(Staff *staff, Doc *doc)
     m_staff = staff;
     if (staff && doc) {
         m_staffHeight = (staff->m_drawingLines - 1) * doc->GetDrawingDoubleUnit(staff->m_drawingStaffSize);
+
+        if (!staff->m_drawingStaffDef)
+            return;
+
+        Object *staffParent = m_staff->m_drawingStaffDef->GetParent();
+        if (!staffParent->Is(STAFFGRP) || staff->GetIdx() == 0) {
+            m_spacingType = SpacingType::System;
+            return;
+        }
+
+        ClassIdsComparison matchType({ STAFFDEF, STAFFGRP });
+        Object *staffChild = staff->m_drawingStaffDef;
+        bool notFirstInGroup = false;
+        while (m_spacingType == SpacingType::None) {
+             ListOfObjects allStaffs;
+             staffParent->FindAllDescendantByComparison(&allStaffs, &matchType, 1);
+             // drop parent from the list of descendant
+             if (!allStaffs.empty() && allStaffs.front() == staffParent) {
+                 allStaffs.erase(allStaffs.begin());
+             }
+             // for first child in staff group parent's symbol should be taken, except
+             // when we had a child which not on the first place in group, than take first symbol
+             notFirstInGroup = notFirstInGroup || (!allStaffs.empty() && allStaffs.front() != staffChild);
+             if (notFirstInGroup) {
+                 StaffGrp *staffGrp = dynamic_cast<StaffGrp *>(staffParent);
+                 if (staffGrp && staffGrp->HasSymbol()) {
+                     switch (staffGrp->GetSymbol()) {
+                         case staffGroupingSym_SYMBOL_brace:
+                             m_spacingType = SpacingType::Brace;
+                             break;
+                         case staffGroupingSym_SYMBOL_bracket:
+                         case staffGroupingSym_SYMBOL_bracketsq:
+                             m_spacingType = SpacingType::Bracket;
+                             break;
+                         default:
+                             m_spacingType = SpacingType::None;
+                     }
+                 }
+             }
+
+             if (m_spacingType == SpacingType::None) {
+                 staffChild = staffParent;
+                 staffParent = staffChild->GetParent();
+                 if (!staffParent || !staffParent->Is(STAFFGRP)) {
+                     m_spacingType = SpacingType::Staff;
+                 }
+             }
+        }
```

Comment:
I wonder if it is necessary to do this for each StaffAlignment. This does not look optimized to me. I think doing it only once for the Doc::m_scoreDef and keeping some sort of a map of the spacingType for each `staff@n` would do the job. What do you think?

Reply:
From my point of view that shouldn't be critical since there is not so big amount of staves in the documents. The code designed to use dynamic_cast only when it's needed and staff groups shouldn't have a lot of childs. 
Also, that would be quite strange that scoreDef holds the map of spacing type above the staff@n.

What if I would store found type in static map inside of StaffAlignment and would have lazy initialization for it. I.e. it will search for type in groups tree only when it's not found in that map and adding it to it?

That would be similar to what you want and placed inside class where it's really used.

Reply:
> From my point of view that shouldn't be critical since there is not so big amount of staves in the documents. The code designed to use dynamic_cast only when it's needed and staff groups shouldn't have a lot of childs.

Yes, but this is not a reason not to put it in a better place. It is also conceptually. 

> Also, that would be quite strange that scoreDef holds the map of spacing type above the staff@n.

I don't agree. We are looking at StaffGrp/StaffDefs. These belong to a ScoreDef, not to a StaffAlignment. It is good to have each StaffAlignment having a m_spacingType member. It is the their calculation that is wrongly placed. 

However, I think there is additional problem with the way it is done now, namely that it does not seem to take into account system optimization (hidden empty staves / staffGrps). So in the end, the best way / place to do it would be from the SystemAligner. This means moving the code to the parent in a method SystemAligner::SetSpacing. There you will be able to check if each staff is actually visible of not. I assume this can change some spacingType values. The call to SystemAligner::SetSpacing it to be done from System::OptimizeScoreDefEnd. This will also be much cleaner.

Reply:
I've submitted commit with changed logic that should work correct with system optimization.
Now I'm struggling with moving these calculations to ScoreDef, by now all my tries and experiments gives only screwed up offsets.

Reply:
My latest proposal was to move it to SytemAligner. Did you try adding a SystemAligner::SetSpacing method to be called on System::OptimizeScoreDefEnd? I cannot see it.